[Windows][Concurrency] Use the same clock as Dispatch.

al45tair opened the pull request:
The Concurrency runtime calculates deadlines for scheduling itself using `swift_get_time()`; unfortunately, on Windows that was using `QueryPerformanceCounter()`, while Dispatch uses
`QueryInterruptTimePrecise()`.  The problem with that is that the two do not necessarily correspond *at all*.  In general
`QueryPerformanceCounter()` may be using any of a number of hardware timers depending on the machine on which we're running.

In the VM I was testing on, the two differed by 20ms, but the worst case is that they are completely unrelated, in which case `Task.sleep()` will wait essentially a random amount of time.

Fixes #72095.

rdar://135413803

al45tair committed:
[Windows][Concurrency] Use the same clock as Dispatch.

The Concurrency runtime calculates deadlines for scheduling itself
using `swift_get_time()`; unfortunately, on Windows that was using
`QueryPerformanceCounter()`, while Dispatch uses
`QueryInterruptTimePrecise()`.  The problem with that is that the two do
not necessarily correspond *at all*.  In general
`QueryPerformanceCounter()` may be using any of a number of hardware
timers depending on the machine on which we're running.

In the VM I was testing on, the two differed by 20ms, but the worst case
is that they are completely unrelated, in which case `Task.sleep()` will
wait essentially a random amount of time.

rdar://135413803

al45tair commented:
@swift-ci Please test

grynspan commented on stdlib/public/Concurrency/Clock.cpp:
```diff
@@ -41,19 +41,36 @@ void swift_get_time(
+      static QueryITP_FP queryITP = nullptr;
+      static swift::once_t onceToken;
+      swift::once(onceToken, [] {
+        if (HMODULE hKernelBase = GetModuleHandleW(L"KernelBase.dll")) {
```
Because you're directly linking mincore.lib, I'm not sure the call to `GetProcAddress()` is needed. The call is needed for `QueryUnbiasedInterruptTimePrecise()` because it's supposed to be in kernel32.lib but appears to have been accidentally omitted.

If you can just call the function directly, there's no need for the `swift::once` call or fallback to `QueryInterruptTime()`.

grynspan replied:
I don't know if it is—we should double-check is all. :)

grynspan replied:
Try updating the PR to just call the function directly—if it's not in the .lib file, it'll fail to build and we'll have our answer.

al45tair replied:
I think we can call it directly; looks like both functions are in `mincore.lib` after all.

grynspan reviewed (changes requested):
You also need to update `swift_get_clock_res()`.

al45tair commented:
> You also need to update `swift_get_clock_res()`.

Nice spot. You're right, I do.

al45tair committed:
[Concurrency][Windows] Also update swift_get_clock_res().

I should have updated this to match as well.

rdar://135413803

al45tair commented:
@swift-ci Please test

al45tair committed:
[Concurrency][Windows] Remove use of `GetProcAddress()`.

Since we're linking `mincore.lib`, we don't need to use
`GetProcAddress()` to find `Query[Unbiased]InterruptTimePrecise()`.

rdar://135413803

al45tair commented:
@swift-ci Please test

grynspan commented on stdlib/public/Concurrency/Clock.cpp:
```diff
@@ -41,19 +41,15 @@ void swift_get_time(
-      continuous.tv_nsec = ns % 1'000'000'000;
+      // This needs to match what swift-corelibs-libdispatch does
+
+      // QueryInterruptTimePrecise() outputs a value measured in 100ns
```
Given how almost-identical this is to the other clock, would it make sense to factor it out and just have a `ULONGLONG_to_timespec()` function?

al45tair replied:
We're only doing this in a couple of places, so I think I'm relaxed about it for now. If we add another one then yes, probably.

jakepetroules commented on stdlib/public/Concurrency/CMakeLists.txt:
```diff
@@ -15,7 +15,10 @@ set(SWIFT_RUNTIME_CONCURRENCY_SWIFT_FLAGS -I${CMAKE_CURRENT_SOURCE_DIR}/Internal
-  list(APPEND swift_concurrency_private_link_libraries Synchronization)
+  list(APPEND swift_concurrency_private_link_libraries
+    Synchronization
+    mincore.lib # For QueryInterruptTime()
```
Not kernel32.lib (per docs)?

al45tair replied:
No. The docs for `QueryInterruptTime()` actually say it's in `mincore.lib`, but the ones for `QueryUnbiasedInterruptTime()` say `kernel32.lib`, however the latter is wrong (I dumped the symbols from `mincore.lib`; sure enough, it has both of them).

grynspan replied:
We should probably report the documentation bug to Microsoft. :) The documentation is open-sourced [here](https://github.com/MicrosoftDocs/sdk-api/blob/b23c70c9092f31a0ff17bb0f60c03fff7855510c/sdk-api-src/content/realtimeapiset/nf-realtimeapiset-queryunbiasedinterrupttimeprecise.md).

compnerd replied:
https://github.com/MicrosoftDocs/sdk-api/pull/1890

grynspan replied:
Thanks!

grynspan commented:
Related: #63225
